IDinsight/ask-a-question
# Added GHA workflow for alembic migrations.

Reviewer: @suzinyou 
Estimate: ~15 min

---

## Ticket

Fixes: [AAQ-759](https://idinsight.atlassian.net/browse/AAQ-759?atlOrigin=eyJpIjoiNzEzYTgwZTNhYWViNGI2NWFhZmY4ZGVjYWIxYWM0MmQiLCJwIjoiaiJ9)

## Description

### Goal

Context

We currently use `pytest-alembic` to test alembic migrations. However, this does not catch issues that occur when there is an open PR against the `main` branch. 

For example, if the open PR branch has not rebased/merged with the latest `main` branch and the `main` branch has a different alembic head (perhaps due to another PR being merged at the time when the open PR is submitted), then `pytest-alembic` will not catch this and the open PR will pass all alembic tests and current workflows. *After* the merge, `pytest-alembic` will catch the multiple heads issue but this is now too late.

Proposed Solution

Add another GHA workflow, `alembic-migrations.yaml`, that will check for outdated alembic heads for open PR's against `main` branch.

This workflow will check for outdated heads DURING the PR and also when the PR is MERGED.

It is necessary to do the check after the PR is merged to account for scenarios when *another* PR is merged with `main` before the open PR in question gets merged. In other words, there can be scenarios where the workflow passes during the PR but then the author delays in getting the PR merged with `main`. During this delay, another PR can be merged with `main`, causing the delayed PR to pass the on `pull_request` step but FAIL after the delayed PR is merged.

Note

The current solution addresses the multiple/outdated heads issue since this is the most egregious one for alembic atm. There are other real-time/live scenarios that `pytest-alembic` might not catch and an updated GHA workflow might be necessary. For such future scenarios, I advise putting the updated logic into `alembic-migrations.yaml`.

### Changes

1. Added `alembic-migrations.yaml`GHA workflow to catch multiple heads issue between open PR branch and current `main` branch. This workflow will trigger during PRs against `main` branch and pushes to `main` branch.

### Future Tasks (optional)

## How has this been tested?

To simulate a test on `pull_request`:

1. Checkout this PR branch.
2. Create `test1` branch from this PR branch. Create `test2` branch from this PR branch.
3. In `test1`, `alembic revision -m "test1" && alembic upgrade head`, commit and push changes to github. Open PR against `tonyzhao6-gha-workflow-for-alembic-migration` and this should pass. Do not merge yet.
4. In `test2` branch, delete the head revision, `alembic revision -m "test2" && alembic upgrade head`, commit and push changes to github. Open PR against `tonyzhao6-gha-workflow-for-alembic-migration` and this should FAIL. Do not merge yet.

Then, to simulate a test on `push`:

1. Fix `test2` branch so that the head history is up-to-date with `tonyzhao6-gha-workflow-for-alembic-migration` branch.  Commit and push changes to github and open PR against `tonyzhao6-gha-workflow-for-alembic-migration` branch. This should pass. Do not merge yet.
2. In PR for `test1` branch, merge with `tonyzhao6-gha-workflow-for-alembic-migration` branch. Wait and verify that the on `push` step passes.
3. In PR for `test2` branch, mrege with `tonyzhao6-gha-workflow-for-alembic-migration` branch. Wait and verify that the on `push` step FAILS. 

Finally,

Clean up fake revisions in `tonyzhao6-gha-workflow-for-alembic-migration` to align with `main` branch for official PR.

## To-do before merge (NOT OPTIONAL)

If all things make sense, I will push one final change to this PR to update the workflow YAML to just activate against `main` branch.

## Checklist

Fill with `x` for completed. 


(Delete any items below that are not relevant)

[AAQ-759]: https://idinsight.atlassian.net/browse/AAQ-759?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
